ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java

```diff
@@ -202,23 +199,68 @@ protected void createThread(ReplicaThread replicaThread, boolean unused) {
 
   /**
    * Assign all replicas of a partition to the same thread to avoid concurrent updates to Azure
-   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo}
-   * @param startThread start threads on creation
+   *
+   * @param partitionId Partition
+   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} List of replicas, must belong to same partition
    */
-  @Override
-  protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean startThread) {
+  protected void addPartitionToReplicaThread(PartitionId partitionId, List<RemoteReplicaInfo> remoteReplicaInfos) {
+    /**
+     * This fn must be called with rwlock.writeLock() held.
+     * It is called by helix when a new partition (not a new replica) is assigned to VCR node.
+     * addPartition -> addPartitionToReplicaThread
+     */
+    // All replicas belong to the same partition, so just assign all to the same thread
+    ReplicaThread rthread = threadPool.get(threadIndex.incrementAndGet() % threadPool.size());
     for (RemoteReplicaInfo rinfo : remoteReplicaInfos) {
       String datacenter = rinfo.getReplicaId().getDataNodeId().getDatacenterName();
       if (!localDatacenterName.equalsIgnoreCase(datacenter)) {
         throw new UnsupportedOperationException(
             String.format("Cross-colo replication from %s to %s is unsupported in VCR for replica {}",
                 localDatacenterName, datacenter, rinfo));
       }
-      ReplicaThread rthread = threadPool.get(threadIndex.incrementAndGet() % threadPool.size());
       rthread.addRemoteReplicaInfo(rinfo);
       rinfo.setReplicaThread(rthread);
-      rthread.startThread();
-      logger.info("Added replica {} to thread {}", rinfo, rthread.getName());
+      logger.info("[PARTITION] Added replica {} to thread {}", rinfo, rthread.getName());
+    }
+    partitionToPartitionInfo.get(partitionId).setReplicaThread(rthread);
+    rthread.startThread();
+  }
+
+  /**
+   * Assign all replicas of a partition to the same thread to avoid concurrent updates to Azure
+   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} These replicas belong to different partitions
+   * @param unused unused variable
+   */
+  @Override
+  protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean unused) {
+    /**
+     * This fn must be called with rwlock.writeLock() held.
+     * It is called by helix when a new replica (not a new partition) is assigned to VCR node.
+     * onReplicaAddedOrRemoved -> addPartitionToReplicaThread
+     */
+    try {
+      for (RemoteReplicaInfo rinfo : remoteReplicaInfos) {
+        try {
+          // Replicas here may belong to different partitions, but they must already be assigned to this node
+          PartitionId partitionId = rinfo.getReplicaId().getPartitionId();
+          ReplicaThread rthread = partitionToPartitionInfo.get(partitionId).getReplicaThread();
```

Comment:
I actually prefer to create a separate map for partition id to thread mapping since it's not the same logic used in server-to-server replication. But if you want to do this, it's fine, please add some comments to the PartitionInfo so we know it's only used for VCR replication.

Reply:
Thanks, I thought using another map is fragmenting the information. We have so many maps in the code.

Reply:
Added a comment.